apps/module-presale/contracts/Presale.sol

```diff
@@ -0,0 +1,336 @@
+        uint256 _percentSupplyOffered,
+        uint64 _fundingPeriod,
+        address _fundraisingPool,
+        address _beneficiaryAddress,
+        uint256 _percentFundingForBenefiriary
+    )
+        external
+        onlyInit
+    {
+        require(isContract(_daiToken), ERROR_INVALID_DAI_TOKEN);
+        require(isContract(_fundraisingPool), ERROR_INVALID_POOL);
+        require(_fundingPeriod > 0, ERROR_INVALID_TIME_PERIOD);
+        require(_vestingCliffPeriod > _fundingPeriod, ERROR_INVALID_TIME_PERIOD);
+        require(_vestingCompletePeriod > _vestingCliffPeriod, ERROR_INVALID_TIME_PERIOD);
+        require(_daiFundingGoal > 0, ERROR_INVALID_DAI_FUNDING_GOAL);
+        require(_percentSupplyOffered > 0, ERROR_INVALID_PERCENT_VALUE);
+        require(_percentSupplyOffered < PPM, ERROR_INVALID_PERCENT_VALUE);
+        require(_beneficiaryAddress != 0x0, ERROR_INVALID_BENEFICIARY_ADDRESS);
+        require(_percentFundingForBenefiriary > 0, ERROR_INVALID_PERCENT_VALUE);
+        require(_percentFundingForBenefiriary < PPM, ERROR_INVALID_PERCENT_VALUE);
+
+        initialized();
+
+        daiToken = _daiToken;
+        _setProjectToken(_projectToken, _projectTokenManager);
+
+        fundraisingPool = _fundraisingPool;
+
+        vestingCliffPeriod = _vestingCliffPeriod;
+        vestingCompletePeriod = _vestingCompletePeriod;
+        fundingPeriod = _fundingPeriod;
+
+        beneficiaryAddress = _beneficiaryAddress;
+        percentFundingForBeneficiary = _percentFundingForBenefiriary;
+
+        daiFundingGoal = _daiFundingGoal;
+        percentSupplyOffered = _percentSupplyOffered;
+
+        _calculateExchangeRate();
+    }
+
+    /*
+     * Public
+     */
+
+    /**
+    * @notice Starts the sale, changing its state to Funding. After the sale is started contributors will be able to purchase project tokens.
+    */
+    function start() public auth(START_ROLE) {
+        require(currentSaleState() == SaleState.Pending, ERROR_INVALID_STATE);
+        startDate = getTimestamp64();
+        emit SaleStarted();
+    }
+
+    /**
+    * @notice Buys project tokens using the provided `_daiToSpend` dai tokens. To calculate how many project tokens will be sold for the provided, dai amount, use daiToProjectTokens(). Each purchase generates a numeric purchaseId (0, 1, 2, etc) for the caller, which can be obtained in the TokensPurchased event emitted, and is required for later refunds.
+    * @param _daiToSpend The amount of dai to spend to obtain project tokens.
+    */
+    function buy(uint256 _daiToSpend) public auth(BUY_ROLE) {
+        require(currentSaleState() == SaleState.Funding, ERROR_INVALID_STATE);
+        require(daiToken.balanceOf(msg.sender) >= _daiToSpend, ERROR_INSUFFICIENT_DAI);
+        require(daiToken.allowance(msg.sender, address(this)) >= _daiToSpend, ERROR_INSUFFICIENT_DAI_ALLOWANCE);
+        require(totalDaiRaised.add(_daiToSpend) <= daiFundingGoal, ERROR_EXCEEDS_FUNDING_GOAL);
```

Comment:
⚠️ This the only behaviour that I strongly disagree with.

Enforcing that the sale can only go to 100%, and that it can only be completed if it reached 100% is incredibly difficult UX.

We could do a number of alternatives:

- Allow the sale to go past 100%, and guard the ability to close the sale early with a role,
- Allow a final contribution to provide more than the amount of tokens left, but do some quick maths to only take as much as needed to get to 100%